src/server/repositories/UserRepository.ts

```diff
@@ -90,39 +93,82 @@ export class UserRepository implements UserRepositoryInterface {
     conditions: UserUrlsQueryConditions,
   ) => Promise<UrlsPaginated> = async (conditions) => {
     const notFoundMessage = 'Urls not found'
-    const userCountAndArray = await User.scope([
+    const { whereConditions, includeConditions } =
+      UserRepository.buildQueryConditions(conditions)
+    const urlsAndCount = await Url.scope([
       'defaultScope',
-      {
-        method: ['urlsWithQueryConditions', conditions],
-      },
+      'getClicks',
     ]).findAndCountAll({
-      subQuery: false, // set limit and offset at end of main query instead of subquery
+      where: whereConditions,
+      distinct: true,
+      limit: conditions.limit,
+      offset: conditions.offset,
+      order: [
+        [
+          { model: UrlClicks, as: 'UrlClicks' },
+          conditions.orderBy,
+          conditions.sortDirection,
+        ],
+      ],
+      include: [includeConditions],
     })
-
-    if (!userCountAndArray) {
+    if (!urlsAndCount) {
       throw new NotFoundError(notFoundMessage)
     }
-
-    const { rows } = userCountAndArray
-    let { count } = userCountAndArray
-    const [userUrls] = rows
-
-    if (!userUrls) {
+    let { rows } = urlsAndCount
+    let { count } = urlsAndCount
+    if (!rows || count === 0) {
       throw new NotFoundError(notFoundMessage)
     }
+    if (conditions.tags && conditions.tags.length > 0) {
+      // Perform a second DB read to retrieve all tags
+      const shortUrls = rows.map((urlType) => {
+        return urlType.shortUrl
+      })
+      rows = await Url.scope()
+        .scope(['defaultScope', 'getClicks', 'getTags'])
+        .findAll({
+          where: { shortUrl: shortUrls },
+        })
+    }
 
-    const urls = userUrls.Urls.map((urlType) =>
-      this.urlMapper.persistenceToDto(urlType),
-    )
-
-    // count will always be >= 1 due to left outer join on user and url tables
-    // to handle edge case where count === 1 but user does not have any urls
+    const urls = rows.map((urlType) => this.urlMapper.persistenceToDto(urlType))
     if (urls.length === 0) {
       count = 0
     }
-
     return { urls, count }
   }
+
+  private static buildQueryConditions(conditions: UserUrlsQueryConditions) {
```

Comment:
Just one small suggestion, can we add an expected output type here so that it's easy to see at one glance what the built query condition object looks like, since there's quite a lot of optional conditions that can go in

Reply:
the function returns [any, any] so it doesn't really add much more information even with the return type. I will do some more research on this but making this optional for now